src/coreclr/binder/inc/defaultassemblybinder.h

```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+#ifndef __DEFAULT_ASSEMBLY_BINDER_H__
+#define __DEFAULT_ASSEMBLY_BINDER_H__
+
+#include "applicationcontext.hpp"
+#include "assemblyloadcontext.h"
+
+class PEAssembly;
+class PEImage;
+
+class DefaultAssemblyBinder final : public AssemblyLoadContext
```

Comment:
Nit: I would prefer `DefaultAssemblyLoadContext`. It just feels weird that the inheritance is `AssemblyBinder` -> `AssemblyLoadContext` -> `DefaultAssemblyBinder`. 
Same for `CustomAssemblyLoadContext`.
But it's just a subjective naming - so feel free to keep as is.

Reply:
I think "Context" works better for something passive - like environment used for binding. When the class itself provides binding functionality it should be Binder, since it is more than just context. 
The base class main purpose is to provide "Bind" APIs, so it feels like it should be "Binder". 

The `AssemblyLoadContext` feels a bit odd in the hierarchy until you realize its purpose is to add to the binding environment a possibility of alternative assembly loader and a link to the managed AssemblyLoadContext counterpart. 

ALC does not add any "Bind" methods and is not normally used for binding (or should not be encouraged). We could in theory imagine a Binder that inherits from AssemblyBinder directly and does not have ALC stuff on it. In theory.

Technically ALC could be merged with the base class, but since it serves distinct purposes, it feels better to keep it a separate concern in the hierarchy and keep its name too.

Reply:
> ALC does not add any "Bind" methods and is not normally used for binding (or should not be encouraged). 

I'm not sure I follow this. Logically `ALC.LoadFromAssemblyPath` is the managed equivalent of `AssemblyBinder.BindUsingPEImage` (this is actually pretty much exactly the relationship). `ALC.LoadFromAssemblyName` 
 maps onto `AssemblyBinder.BindUsingAssemblyName` (it's also a pretty good match in reality). It's not exact since there are additional layers between the managed code and native (to deal with marshalling validation and so on). So while I agree that nobody in the runtime should call the Bind methods directly, they are sort of called "directly" from managed a lot.

But I guess I see your point. Runtime has a separate concept of "binding" (finding the right file for a given assembly spec) and a separate concept of "loading" (actually parsing a file and making it available to run code from). The managed APIs don't have these as separate concepts (although in the past we discussed that having it would be beneficial in some cases).

I still think that the sealed classes should be called `DefaultAssemblyLoadContext`/`CustomAssemblyLoadContext` - since they're a specialization of `AssemblyLoadContext` - but it's just a subjective naming thing :-)

Reply:
`DefaultAssemblyLoadContext`/`CustomAssemblyLoadContext` would fit in my brain better as well. Although I also have the native `AssemblyLoadContext` and `AssemblyBinder` classes lumped together in my brain, so that may be why.

> it serves distinct purposes, it feels better to keep it a separate concern in the hierarchy and keep its name too.

Since it is so subjective, it might help to have comments in the headers with the intent / describing the intentional separation.

Reply:
As we discuss this I am getting more and more convinced that native `AssemblyLoadContext` should probably be merged with its base. 

Native `AssemblyLoadContext` is not a native representation of the managed one, it just _has_ a link to the managed, and can provide it via `GetManagedAssemblyLoadContext`. As a matter of completely mechanical naming it would be something like `AssemblyBinderWithManagedAssemblyLoadContext`.

However we do not really have concrete binders that cannot provide a managed context. If we do not envision such a case, then maybe we do not need an abstract one either.
Native `AssemblyLoadContext` seems to add more naming confusion than actual functionality (which is not all that much).

Reply:
> Native AssemblyLoadContext is not a native representation of the managed one, it just has a link to the managed, and can provide it via GetManagedAssemblyLoadContext

True, but for all intents and purposes `AssemblyLoadContext`-native has 1:1 relationship with `AssemblyLoadContext`-managed. The native one has a pointer to the managed one (and vice versa). Also all of the functionality of managed ALC is implemented (indirectly) in the native ALC (lot of this is hidden in the `ApplicationContext` which is embeded in the native ALC - and yes - that name is TERRIBLE and has to go :-), but that's a different refactoring)

Reply:
> the `ApplicationContext` which is embeded in the native ALC - and yes - that name is TERRIBLE and has to go

The `ApplicationContext` which also contains an `ExecutionContext` which is a subclass of  `LoadContext`! So much context.

Reply:
And the fact that in managed there is a public class `ApplicationContext` (in [winforms](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.applicationcontext?view=net-5.0)) and that there is an `AppContext` class (also public) in [CoreLib](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext?view=net-5.0#:~:text=The%20AppContext%20class%20enables%20library%20writers%20to%20provide,when%20a%20change%20is%20made%20to%20existing%20functionality.). Either of which has absolutely no relationship to the native one :-)